grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -352,4 +360,75 @@ class PrepaymentInvoiceService {
 
         return quantityInvoiced >= quantityInverseable ? quantityInverseable : quantityInvoiced
     }
+
+    /**
+     * Updates invoice item's unit price and amount and finds related inverse item and updates unit price and amount
+     * */
+    void updateInvoiceItemUnitPrice(String itemId, BigDecimal unitPrice) {
+        def g = grailsApplication.mainContext.getBean('org.grails.plugins.web.taglib.ApplicationTagLib')
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (!invoiceItem) {
+            String defaultMessage = "Cannot find invoice item with id: ${itemId}"
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.cannotFind.error", args: [itemId], default: defaultMessage))
+        }
+        if (invoiceItem.invoice.datePosted) {
+            String defaultMessage = "Cannot update posted invoices"
+            throw new IllegalArgumentException(g.message(code: "invoice.cannotUpdatePosted.error", default: defaultMessage))
+        }
+        if (unitPrice == 0) {
+            String defaultMessage = "Cannot update unit price to 0"
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.unitPriceZero.error", default: defaultMessage))
+        }
+        if (invoiceItem?.inverse) {
+            String defaultMessage = "Cannot edit inverse invoice items directly"
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.cannotEditInverse.error", default: defaultMessage))
+        }
+        if (!invoiceItem?.orderAdjustment) {
+            String defaultMessage = "Invoice item is missing order adjustment. Cannot edit unit price."
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.missingAdjustment.error", default: defaultMessage))
+        }
+        BigDecimal totalAdjustments = invoiceItem.orderAdjustment?.totalAdjustments
+        if ((totalAdjustments < 0 && unitPrice > 0) || (totalAdjustments > 0 && unitPrice < 0)) {
+            String defaultMessage = "Cannot change the positive unit price to negative or negative to positive."
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.unitPriceSignChange.error", default: defaultMessage))
+        }
+
+        BigDecimal amountAvailableToInvoice = invoiceItem.orderAdjustment.amountAvailableToInvoice
+        if (Math.abs(unitPrice) > Math.abs(amountAvailableToInvoice) + Math.abs(invoiceItem.unitPrice)) {
+            String defaultMessage = "Cannot update unit price to higher value than available to invoice"
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.unitPriceTooHigh.error", default: defaultMessage))
+        }
+        // update invoice item's unit price (and adjust amount)
+        invoiceItem.unitPrice = unitPrice
+        invoiceItem.amount = invoiceItem.quantity * unitPrice
```

Comment:
this is fine, but if we ever forget that amount is derived from quantity and unit price, I can imagine a scenario where it falls out of sync. Maybe it would be better to have a custom method that auto-recomputes amount when a new unit price or quantity is set.